pkg/core/statesync/module.go

```diff
@@ -429,6 +512,81 @@ func (s *Module) AddMPTNodes(nodes [][]byte) error {
 	return nil
 }
 
+// AddContractStorageData adds a batch of key-value pairs for storage-based sync.
+func (s *Module) AddContractStorageData(kvs []KeyValue, syncHeight uint32, expectedRoot util.Uint256) error {
+	if s.mode == MPTBased {
+		panic("contract storage items are not expected in MPT-based mode")
+	}
+	oldStage := s.syncStage
+	s.lock.Lock()
+	defer func() {
+		if s.syncStage != oldStage {
+			s.notifyStageChanged()
+		}
+	}()
+	defer s.lock.Unlock()
+
+	if s.syncStage&headersSynced == 0 || s.syncStage&mptSynced != 0 || expectedRoot.Equals(s.localTrie.StateRoot()) {
+		return errors.New("contract storage items were not requested")
+	}
+	if syncHeight != s.syncPoint {
+		return fmt.Errorf("invalid sync height: got %d, expected %d", syncHeight, s.syncPoint)
+	}
+
+	var (
+		prefix  = TemporaryPrefix(s.dao.Version.StoragePrefix)
+		lastKey []byte
+		batch   = make(map[string][]byte, len(kvs))
+	)
+	for _, kv := range kvs {
+		s.dao.Store.Put(append([]byte{byte(prefix)}, kv.Key...), kv.Value)
+		batch[string(kv.Key)] = kv.Value
+		lastKey = kv.Key
+	}
+	if _, err := s.dao.Store.PersistSync(); err != nil {
+		return fmt.Errorf("failed to persist contract storage data: %w", err)
+	}
+
+	if len(batch) > 0 {
```

Comment:
The case of zero-length batch should be handled in the beginning of the function.